### ISSUE TITLE
Deleting an entity

### ISSUE THREAD

**Issue by FireFlyForLife** [VIGameStudio/Concordia#3]
The problem: When deleting an entity, all you have is its ID. so there is no way of knowing which components you have to delete.

My solution: I added a base ICmpPool class which knows the sizeof the the type and the starting byte*. Along with a vector which holds all the entity ids. the index of the array can be directly translated to the offset of the component (`byte* + element_size * index`). wherer the data can be erased (when deleting).

Cons (Apart from type safety): just erasing the data does not call the deconstructor of the classes that are stored.
I also do not know how well this works for complex classes. For example, I don't know how well it plays with std::functions and their magic.

I see no other way doing it, since I think the `static size_t counter` gets initialized at runtime.

**Comment by FireFlyForLife** [comment]
I made a little snippet to illustrate what I mean. As you can see from the error at line 12, the function needs to be constexpr to be used at compile time, however that is not possible because... its not constant. 

cpp.sh/4guer

**Comment by Belfer** [comment]
The point of the component pool was to associate a component with an entity ID. We can use this to locate the corresponding component index in the pool and disable it for reuse.

I'm not sure I understand what it is you are trying to accomplish with the example code?

**Comment by FireFlyForLife** [comment]
However this requires you to have the type of the component, which you do not have in a situation when you delete a entity, since you only have the entity_id which is a number. so you cannot do getPool<C>() because you do not have the C type. so you need to have a way of accessing the pools to check if they contain that entityID. Which is why I created that std::vector in the base ICmpPool.

**Comment by Belfer** [comment]
So it's rather more to do with disassociating the components an entity has when deleting it?

**Comment by FireFlyForLife** [comment]
Yes like I said, you can add a component with the type and get the ID from the type, but not get the type from ID which is the problem.

**Comment by Belfer** [comment]
I think with some modifications to the pool class we should be able to work around this issue. For one we could directly map the component ID's as a index to an std::vector of pool*. This means that we could have access to the pool vector without the need of templates. We could then modify the pool interface to expose the methods of the pool class and use that to blindly find the component that matches the entity ID and disable it.

**Comment by Belfer** [comment]
All we need to do then is simply keep track of the component ID's the entity has.

**Comment by FireFlyForLife** [comment]
Thats pretty much the [ICmpPool ](https://github.com/FireFlyForLife/Concordia/blob/master/src/Concordia/ComponentPool.hpp)class I wrote.

**Comment by Belfer** [comment]
There's no need to rewrite that class as it's pretty standard and generic, exposing the binding calls to std::vector (add, remove, etc) should suffice.

**Comment by Belfer** [comment]
And I also don't think that's what I meant.

**Comment by FireFlyForLife** [comment]
How do you want to expose those calls? have it be a virtual function which the templated Pool implements?

**Comment by Belfer** [comment]
Yes, indeed.

**Comment by Belfer** [comment]
I can leave this to you if you'd like to do it? Otherwise I'll do it when I have some time.

**Comment by FireFlyForLife** [comment]
Yea sure no problem!

**Comment by Belfer** [comment]
Excellent! Make sure you reference this issue when committing a change relating to this, and also don't forget to use this to post any questions or suggestions on the design.
